File: Nexus/Source/DefinitionsTests/CountryTester.cpp
#include <Beam/SerializationTests/ValueShuttleTests.hpp>
#include <Beam/Utilities/ToString.hpp>
#include <doctest/doctest.h>
#include "Nexus/Definitions/Country.hpp"

using namespace Beam;
using namespace Beam::Tests;
using namespace Nexus;

TEST_SUITE("Country") {
  TEST_CASE("comparison") {
    auto default_code = CountryCode();
    REQUIRE(!default_code);
    auto code1 = CountryCode(42);
    auto code2 = CountryCode(42);
    auto code3 = CountryCode(7);
    REQUIRE(code1 == code2);
    REQUIRE(code3 < code1);
    REQUIRE(code1 > code3);
  }

  TEST_CASE("empty") {
    REQUIRE_FALSE(static_cast<bool>(CountryCode()));
    REQUIRE(static_cast<bool>(CountryCode(840)));
  }

  TEST_CASE("stream") {
    auto code = CountryCode(156);
    REQUIRE(to_string(code) == "CN");
    test_round_trip_shuttle(code);
  }

  TEST_CASE("stream_unknown") {
    REQUIRE(to_string(CountryCode(9999)) == "9999");
  }

  TEST_CASE("context_stream") {
    auto entry1 = CountryDatabase::Entry();
    entry1.m_code = CountryCode(124);
    entry1.m_name = "Canada";
    entry1.m_two_letter_code = "CA";
    entry1.m_three_letter_code = "CAN";
    auto database = CountryDatabase();
    database.add(entry1);
    auto ss = std::stringstream();
    ss << database << entry1.m_code;
    REQUIRE(ss.str() == "CA");
  }

  TEST_CASE("none") {
    auto none = CountryDatabase::NONE;
    REQUIRE(!none.m_code);
    REQUIRE(none.m_name == "");
    REQUIRE(none.m_two_letter_code == "??");
    REQUIRE(none.m_three_letter_code == "???");
  }

  TEST_CASE("code_lookup") {
    auto entry1 = CountryDatabase::Entry();
    entry1.m_code = CountryCode(2);
    entry1.m_name = "Two";
    entry1.m_two_letter_code = "TT";
    entry1.m_three_letter_code = "TTT";
    auto database = CountryDatabase();
    database.add(entry1);
    auto entry2 = CountryDatabase::Entry();
    entry2.m_code = CountryCode(1);
    entry2.m_name = "One";
    entry2.m_two_letter_code = "OO";
    entry2.m_three_letter_code = "OOO";
    database.add(entry2);
    auto entries = database.get_entries();
    REQUIRE(entries.size() == 2);
    REQUIRE(entries[0].m_code == CountryCode(1));
    REQUIRE(entries[1].m_code == CountryCode(2));
    REQUIRE(database.from(CountryCode(1)).m_name == "One");
    REQUIRE(!database.from(CountryCode(3)).m_code);
  }

  TEST_CASE("name_lookup") {
    auto database = CountryDatabase();
    auto entry = CountryDatabase::Entry();
    entry.m_code = CountryCode(5);
    entry.m_name = "FiveLand";
    entry.m_two_letter_code = "FL";
    entry.m_three_letter_code = "FIV";
    database.add(entry);
    REQUIRE(database.from_name("FiveLand").m_code == CountryCode(5));
    REQUIRE(database.from("FL").m_code == CountryCode(5));
    REQUIRE(database.from("FIV").m_code == CountryCode(5));
    REQUIRE(!database.from_name("Unknown").m_code);
  }

  TEST_CASE("remove") {
    auto database = CountryDatabase();
    auto entry = CountryDatabase::Entry();
    entry.m_code = CountryCode(9);
    entry.m_name = "Nine";
    entry.m_two_letter_code = "NI";
    entry.m_three_letter_code = "NIN";
    database.add(entry);
    REQUIRE(database.from(CountryCode(9)).m_code == CountryCode(9));
    database.remove(CountryCode(9));
    REQUIRE(!database.from(CountryCode(9)).m_code);
  }

  TEST_CASE("shuttle") {
    test_round_trip_shuttle(DEFAULT_COUNTRIES, [] (const auto& countries) {
      auto expected_entries = DEFAULT_COUNTRIES.get_entries();
      auto entries = countries.get_entries();
      REQUIRE(expected_entries.size() == entries.size());
      for(auto i = std::size_t(0); i != entries.size(); ++i) {
        REQUIRE(expected_entries[i] == entries[i]);
      }
    });
  }

  TEST_CASE("parse") {
    auto database = CountryDatabase();
    auto entry = CountryDatabase::Entry();
    entry.m_code = CountryCode(8);
    entry.m_name = "Eight";
    entry.m_two_letter_code = "EI";
    entry.m_three_letter_code = "EGT";
    database.add(entry);
    REQUIRE(parse_country_code("EI", database) == CountryCode(8));
    REQUIRE(parse_country_code("EGT", database) == CountryCode(8));
    REQUIRE(parse_country_code("Eight", database) == CountryCode(8));
    REQUIRE(!parse_country_code("Unknown", database));
  }


  TEST_CASE("parse_country_database_entry") {
    auto yaml = R"(
      - name: "United States"
        two_letter_code: "US"
        three_letter_code: "USA"
        code: 840)";
    auto node = YAML::Load(yaml);
    auto entry = parse_country_database_entry(node[0]);
    REQUIRE(entry.m_name == "United States");
    REQUIRE(entry.m_two_letter_code == "US");
    REQUIRE(entry.m_three_letter_code == "USA");
    REQUIRE(entry.m_code == CountryCode(840));
  }

  TEST_CASE("parse_country_database") {
    auto yaml = R"(
    - name: "Country A"
      two_letter_code: "AA"
      three_letter_code: "AAA"
      code: 10
    - name: "Country B"
      two_letter_code: "BB"
      three_letter_code: "BBB"
      code: 20)";
    auto node = YAML::Load(yaml);
    auto database = parse_country_database(node);
    REQUIRE(database.get_entries().size() == 2);
    auto country_a = database.from("AA");
    REQUIRE(country_a.m_name == "Country A");
    REQUIRE(country_a.m_two_letter_code == "AA");
    REQUIRE(country_a.m_three_letter_code == "AAA");
    REQUIRE(country_a.m_code == CountryCode(10));
    auto country_b = database.from("BB");
    REQUIRE(country_b.m_name == "Country B");
    REQUIRE(country_b.m_two_letter_code == "BB");
    REQUIRE(country_b.m_three_letter_code == "BBB");
    REQUIRE(country_b.m_code == CountryCode(20));
  }
}

File: WebApi/tests/definitions/country_tester.ts
import { describe, it } from 'node:test';
import * as assert from 'node:assert';
import { CountryCode, CountryDatabase } from '../../source/definitions/country';

describe('CountryCode', () => {
  it('construct', () => {
    const code = new CountryCode(840);
    assert.strictEqual(code.code, 840);
  });

  it('none', () => {
    assert.strictEqual(CountryCode.NONE.code, 65535);
  });

  it('equals', () => {
    const a = new CountryCode(42);
    const b = new CountryCode(42);
    assert.ok(a.equals(b));
  });

  it('not_equal', () => {
    const a = new CountryCode(42);
    const b = new CountryCode(7);
    assert.ok(!a.equals(b));
  });

  it('to_json', () => {
    const code = new CountryCode(840);
    assert.strictEqual(code.toJson(), 840);
  });

  it('from_json', () => {
    const code = CountryCode.fromJson(840);
    assert.strictEqual(code.code, 840);
  });

  it('round_trip_json', () => {
    const original = new CountryCode(124);
    const restored = CountryCode.fromJson(original.toJson());
    assert.ok(original.equals(restored));
  });

  it('hash', () => {
    const a = new CountryCode(840);
    const b = new CountryCode(840);
    assert.strictEqual(a.hash(), b.hash());
  });
});

describe('CountryDatabase', () => {
  it('none_entry', () => {
    const none = CountryDatabase.Entry.NONE;
    assert.ok(none.code.equals(CountryCode.NONE));
    assert.strictEqual(none.name, 'None');
    assert.strictEqual(none.twoLetterCode, '??');
    assert.strictEqual(none.threeLetterCode, '???');
  });

  it('code_lookup', () => {
    const database = new CountryDatabase();
    database.add(new CountryDatabase.Entry(
      new CountryCode(2), 'Two', 'TT', 'TTT'));
    database.add(new CountryDatabase.Entry(
      new CountryCode(1), 'One', 'OO', 'OOO'));
    assert.strictEqual(database.fromCode(new CountryCode(1)).name, 'One');
    assert.strictEqual(database.fromCode(new CountryCode(2)).name, 'Two');
    assert.ok(
      database.fromCode(new CountryCode(3)).code.equals(CountryCode.NONE));
  });

  it('letter_code_lookup', () => {
    const database = new CountryDatabase();
    database.add(new CountryDatabase.Entry(
      new CountryCode(5), 'FiveLand', 'FL', 'FIV'));
    assert.ok(
      database.fromLetterCode('FL').code.equals(new CountryCode(5)));
    assert.ok(
      database.fromLetterCode('FIV').code.equals(new CountryCode(5)));
    assert.ok(
      database.fromLetterCode('XX').code.equals(CountryCode.NONE));
  });

  it('remove', () => {
    const database = new CountryDatabase();
    database.add(new CountryDatabase.Entry(
      new CountryCode(9), 'Nine', 'NI', 'NIN'));
    assert.ok(
      database.fromCode(new CountryCode(9)).code.equals(new CountryCode(9)));
    database.delete(new CountryCode(9));
    assert.ok(
      database.fromCode(new CountryCode(9)).code.equals(CountryCode.NONE));
  });

  it('add_duplicate_ignored', () => {
    const database = new CountryDatabase();
    database.add(new CountryDatabase.Entry(
      new CountryCode(1), 'One', 'OO', 'OOO'));
    database.add(new CountryDatabase.Entry(
      new CountryCode(1), 'Duplicate', 'DD', 'DDD'));
    assert.strictEqual(database.fromCode(new CountryCode(1)).name, 'One');
  });

  it('entry_equals', () => {
    const a = new CountryDatabase.Entry(
      new CountryCode(840), 'United States', 'US', 'USA');
    const b = new CountryDatabase.Entry(
      new CountryCode(840), 'United States', 'US', 'USA');
    assert.ok(a.equals(b));
  });

  it('entry_not_equal', () => {
    const a = new CountryDatabase.Entry(
      new CountryCode(840), 'United States', 'US', 'USA');
    const b = new CountryDatabase.Entry(
      new CountryCode(124), 'Canada', 'CA', 'CAN');
    assert.ok(!a.equals(b));
  });

  it('entry_to_json', () => {
    const entry = new CountryDatabase.Entry(
      new CountryCode(840), 'United States', 'US', 'USA');
    const json = entry.toJson();
    assert.strictEqual(json.code, 840);
    assert.strictEqual(json.name, 'United States');
    assert.strictEqual(json.two_letter_code, 'US');
    assert.strictEqual(json.three_letter_code, 'USA');
  });

  it('entry_from_json', () => {
    const json = {
      code: 840,
      name: 'United States',
      two_letter_code: 'US',
      three_letter_code: 'USA'
    };
    const entry = CountryDatabase.Entry.fromJson(json);
    assert.ok(entry.code.equals(new CountryCode(840)));
    assert.strictEqual(entry.name, 'United States');
    assert.strictEqual(entry.twoLetterCode, 'US');
    assert.strictEqual(entry.threeLetterCode, 'USA');
  });

  it('round_trip_json', () => {
    const database = new CountryDatabase();
    database.add(new CountryDatabase.Entry(
      new CountryCode(840), 'United States', 'US', 'USA'));
    database.add(new CountryDatabase.Entry(
      new CountryCode(124), 'Canada', 'CA', 'CAN'));
    const json = { entries: database.toJson() };
    const restored = CountryDatabase.fromJson(json);
    assert.ok(
      restored.fromCode(new CountryCode(840)).name === 'United States');
    assert.ok(restored.fromCode(new CountryCode(124)).name === 'Canada');
  });

  it('iterable', () => {
    const database = new CountryDatabase();
    database.add(new CountryDatabase.Entry(
      new CountryCode(1), 'One', 'OO', 'OOO'));
    database.add(new CountryDatabase.Entry(
      new CountryCode(2), 'Two', 'TT', 'TTT'));
    const entries = [...database];
    assert.strictEqual(entries.length, 2);
  });
});
